### PR TITLE
Replace NSTimer runLoop with scheduled commands on main thread using GCD

### PR DESCRIPTION
**Pull request by johannth**
There were slight issues with our current NSTimer approach. Replacing the timer with scheduling tasks on the main queue should have the same effect in a more robust and probably more efficient way.

@alliannas this will probably conflict slightly with your PR so I recommend you take a look at this before continuing with your PR.

### CODE REVIEW

File: PVGTableViewProxy/PVGTableViewProxy.m [hunk 1]
```diff
@@ -157,13 +150,15 @@ - (void)addSection:(PVGTableViewSection *)section atIndex:(NSInteger)sectionInde
...
-        self.pendingRenderCommand = RACTuplePack(@(sectionIndex), [newViewModels copy]);
+        [self scheduleRenderCommand:[PVGTableViewRenderCommand renderCommandForSection:sectionIndex
+                                                                            viewModels:[newViewModels copy]]];
     }];
     
     [[section.dataSource.scrollCommands ignore:nil] subscribeNext:^(PVGTableViewScrollCommand *command) {
         @strongify(self);
         BOOL success = [self scrollInSection:sectionIndex usingCommand:command];
         self.pendingScrollCommand = success ? nil : RACTuplePack(@(sectionIndex), command);
+        // FIXME: Do we need to schedule a render here?
```
Comment (johannth) [inline PVGTableViewProxy/PVGTableViewProxy.m]: Preliminary tests say no, we don't need it :)

File: PVGTableViewProxy/PVGTableViewRenderCommand.m [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+@property (readwrite, nonatomic) NSArray<NSObject <PVGTableViewCellViewModel> *> *viewModels;
+
+@end
+
+@implementation PVGTableViewRenderCommand
+
++ (instancetype)renderCommandForSection:(NSInteger)sectionIndex
+                             viewModels:(NSArray<NSObject <PVGTableViewCellViewModel> *> *)viewModels
+{
+    PVGTableViewRenderCommand *command = [PVGTableViewRenderCommand new];
```
Comment (alliannas) [inline PVGTableViewProxy/PVGTableViewRenderCommand.m]: Shouldn't this be `[self new]` for correct behavior in sublcasses?

File: PVGTableViewProxy/PVGTableViewRenderCommand.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+//
+//
+
+#import <Foundation/Foundation.h>
+#import "PVGTableViewCellViewModel.h"
+
+@interface PVGTableViewRenderCommand : NSObject
+
+@property (readonly, nonatomic) NSInteger sectionIndex;
+@property (readonly, nonatomic) NSArray<NSObject <PVGTableViewCellViewModel> *> *viewModels;
```
Comment (alliannas) [inline PVGTableViewProxy/PVGTableViewRenderCommand.h]: I like the extra spaces for the category, makes it bit more readable :+1:

File: Tests/Tests/TableViewProxyTests.m [hunk 4]
```diff
@@ -418,7 +408,7 @@ - (void)test_scroll_in_section_handles_nil_command
     
     [[[self.mockTableView reject] ignoringNonObjectArgs] scrollToRowAtIndexPath:OCMOCK_ANY atScrollPosition:0 animated:NO];
     
-    XCTAssertEqual([self.proxy scrollInSection:0 usingCommand:nil], NO);
+    XCTAssertEqual([self.proxy scrollInSection:0 usingCommand:nil], YES);
```
Comment (alliannas) [inline Tests/Tests/TableViewProxyTests.m]: Minor nitpicking: Should the name of the tests could be more descriptive?
E.g. `test_scroll_in_section_returns_yes_when_command_is_nil`

File: PVGTableViewProxy/PVGTableViewRenderCommand.m (hunk 2 above)
Comment (johannth) [inline PVGTableViewProxy/PVGTableViewRenderCommand.m]: :+1: Good spot.

File: Tests/Tests/TableViewProxyTests.m (hunk 4 above)
Comment (johannth) [inline Tests/Tests/TableViewProxyTests.m]: :+1:

Comment (johannth) [inline Tests/Tests/TableViewProxyTests.m]: Fixed in #	b628b25

**Comment by johannth** [comment]
I'm doing some preliminary testing on this still. I think I'll add in a extra commit to disable the logging (because it affects performance drastically - NSLog boooo)

File: PVGTableViewProxy/PVGTableViewProxy.m [hunk 5]
```diff
@@ -86,31 +95,30 @@ - (instancetype)initWithTableView:(UITableView *)tableView
...
-- (NSTimer *)runLoopTimer
+- (void)scheduleRenderCommand:(PVGTableViewRenderCommand *)renderCommand
 {
-    NSRunLoop *mainLoop = [NSRunLoop mainRunLoop];
-    NSTimer *timer = [NSTimer scheduledTimerWithTimeInterval:0.1 target:self selector:@selector(runLoop) userInfo:nil repeats:YES];
-    [mainLoop addTimer:timer forMode:NSRunLoopCommonModes];
-    
-    return [NSTimer scheduledTimerWithTimeInterval:0.1 target:self selector:@selector(runLoop) userInfo:nil repeats:YES];
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self executeRenderCommand:renderCommand];
```
Comment (alliannas) [inline PVGTableViewProxy/PVGTableViewProxy.m]: Should this be a weak reference?

Comment (johannth) [inline PVGTableViewProxy/PVGTableViewProxy.m]: Hmm, I though about it but I don't think so - http://stackoverflow.com/questions/21987067/using-weak-self-in-dispatch-async-function

... unless we want to make sure that we don't prolong the lifetime of the proxy by a last render. What do you think?

Comment (alliannas) [inline PVGTableViewProxy/PVGTableViewProxy.m]: I think this is fine as is :) Thanks for the so post, very informative

**Comment by johannth** [comment]
@alliannas I think this is ready. I also think it removes the need for your fix (but your section example is invaluable). Could you have another look and then we'd release a new version?

**Comment by alliannas** [comment]
Yes I agree, my fix is out dated (damn mobile industry moves fast :scream:)

I'll probably just close my existing PR and create a new one with a section example, once this is in.

**Comment by johannth** [comment]
I have no idea what is failing on the push job on Travis. Do you @alliannas ?

**Comment by alliannas** [comment]
No :( They have been flaky as hell for some reason.

**Pull request merged**